# Real-world Data Wrangling

In this project, you will apply the skills you acquired in the course to gather and wrangle real-world data with two datasets of your choice.

You will retrieve and extract the data, assess the data programmatically and visually, accross elements of data quality and structure, and implement a cleaning strategy for the data. You will then store the updated data into your selected database/data store, combine the data, and answer a research question with the datasets.

Throughout the process, you are expected to:

1. Explain your decisions towards methods used for gathering, assessing, cleaning, storing, and answering the research question
2. Write code comments so your code is more readable

Before you start, install the some of the required packages. 

In [372]:
!python -m pip install kaggle==1.6.12

In [ ]:
!pip install --target=/workspace ucimlrepo numpy==1.24.3

**Note:** Restart the kernel to use updated package(s).

## 1. Gather data

In this section, you will extract data using two different data gathering methods and combine the data. Use at least two different types of data-gathering methods.

### **1.1.** Problem Statement
In 2-4 sentences, explain the kind of problem you want to look at and the datasets you will be wrangling for this project.

Finding the right datasets can be time-consuming. Here we provide you with a list of websites to start with. But we encourage you to explore more websites and find the data that interests you.

* Google Dataset Search https://datasetsearch.research.google.com/
* The U.S. Government’s open data https://data.gov/
* UCI Machine Learning Repository https://archive.ics.uci.edu/ml/index.php


There are many factors that cause pollution. Some factors are well-known while other factors are not. One factor that is not widely known is population. The project will be addressing the following research question: **Do the most polluted cities have the largest populations?**

### **1.2.** Gather at least two datasets using two different data gathering methods

List of data gathering methods:

- Download data manually
- Programmatically downloading files
- Gather data by accessing APIs
- Gather and extract data from HTML files using BeautifulSoup
- Extract data from a SQL database

Each dataset must have at least two variables, and have greater than 500 data samples within each dataset.

For each dataset, briefly describe why you picked the dataset and the gathering method (2-3 full sentences), including the names and significance of the variables in the dataset. Show your work (e.g., if using an API to download the data, please include a snippet of your code). 

Load the dataset programmtically into this notebook.

#### **Dataset 1**

**_Type_**: CSV File

**_Method_**: The data was gathered using the "_Downloading data manually_" method from the **"edit_AIR QUALITY INDEX (by cities) - IQAir"** CSV file.

The method of "_Downloading data manually_" was used to gather the data from the first dataset because it is the simplest and easiest method to use to gather data from a dataset while keeping an original and edited version of the dataset out of all five methods. The method requires the file of the dataset to be downloaded into the location where all the project files are being held first. Finally, only one line of code is used to read and access the data from the file of the dataset. To keep an original and edited version of the dataset, the dataset simply needs to be downloaded twice and kept under different names in the location where all the project files are being held. This shows that the method of "_Downloading data manually_" is the simplest and easiest method out of all five methods.  

**_Dataset variables_**:

*   *_**"2019"**_: Air Quality Index (AQI) values for measuring Pollution during 2019*
*   *_**"City"**_: Both Country and City names*

The **"2019"** column is needed because it contains the air quality index values that measure the pollution of different cities during 2019. The air quality index values during 2019 are the only air quality index values that can be shared with the other dataset because the other dataset only has values from 2019. These values are also needed to answer the research question since they have values of pollution.

The **"City"** column is needed because the data in the column will be used to merge the two datasets together after the **city** names are put into a new column that is separate from the column containing the **country** names.

The rest of the columns containing _rank_ and the air quality index values for each year from _2017_ to _2021_, except _2019_, are not needed to answer the research question.

In [403]:
#FILL IN 1st data gathering and loading method
import pandas as pd
import numpy as np

df_polluted_cities_and_countries = pd.read_csv("edit_AIR QUALITY INDEX (by cities) - IQAir.csv") # loads CSV file of first dataset
df_polluted_cities_and_countries # displays first dataset

Rank                   City   2021 JAN(2021) FEB(2021) MAR(2021)  \
0        1         Bhiwadi, India  106.2     145.8     129.8     120.2   
1        2       Ghaziabad, India  102.0     199.9     172.2      97.8   
2        3           Hotan, China  101.5         -         -       158   
3        4           Delhi, India   96.4     183.7     142.2      80.5   
4        5         Jaunpur, India   95.3     182.2     143.5        91   
...    ...                    ...    ...       ...       ...       ...   
6470  6471  Mornington, Australia    2.4         2       1.9       2.3   
6471  6472   Emu River, Australia    2.1       1.9       1.8         2   
6472  6473     Judbury, Australia    2.0       1.6       1.5       2.1   
6473  6474   St Helens, Australia    1.9       1.8       2.1         2   
6474  6475        Chu, Kazakhstan    1.5       1.5       1.6       1.6   

     APR(2021) MAY(2021) JUN(2021) JUL(2021) AUG(2021) SEP(2021) OCT(2021)  \
0        125.7      86.5      95.9      55.6      55.4      37.1      91.1   
1         86.3      52.9      47.2      35.3      37.6      30.8      89.7   
2         91.1     167.4      57.4      70.9      93.2      79.3     126.1   
3         72.9      47.4      47.1      35.6      36.9      30.2      73.7   
4           70      51.1      40.7      33.5      34.2      36.8      75.7   
...        ...       ...       ...       ...       ...       ...       ...   
6470       2.1       3.2       3.6       4.3       2.1       2.1       1.7   
6471       2.6       3.4       2.6       1.9       2.1       2.2       1.5   
6472       1.5       4.1         2       2.2       2.2       1.7       1.5   
6473       2.4       2.7       1.6       1.6       1.6       1.9       1.6   
6474       1.5       1.5       1.5       1.5       1.6       1.5       1.5   

     NOV(2021) DEC(2021)   2020   2019   2018   2017  
0        188.6     136.6   95.5   83.4  125.4      -  
1        218.3       163  106.6  110.2  135.2  144.6  
2        111.5      62.6  110.2  110.1    116   91.9  
3        224.1     186.4   84.1   98.6  113.5  108.2  
4          196     195.7      -      -      -      -  
...        ...       ...    ...    ...    ...    ...  
6470         2       1.9    3.2    3.8      3    3.9  
6471       1.4       1.5    2.6    2.5    2.6    2.3  
6472       1.4       1.7    2.4    5.7    2.2    1.9  
6473       2.4       1.6    2.4    2.4    2.9    3.3  
6474       1.5       1.5      -      -      -      -  

[6475 rows x 19 columns]

#### Dataset 2

**_Type_**: CSV File

**_Method_**: The data was gathered using the "Programmatically download files" method from the **"worldcities"** CSV file.

The method of "_Programmatically download files_" was used to gather the data from the second dataset because it is the next simplest and easiest method to use to gather data from a dataset 
out of all five methods. The method first requires the following import to be included in the jupyter notebook file: **import zipfile**. Next, the zip file of the dataset needs to be downloaded into the location where all the project files are being held. Next, two lines of code are used to unzip the zip file in read mode. Finally, the data from the file of the second dataset in the zip file is accessed using the same gathering method used for the first dataset, except the directory for the CSV file is used in the gethering method instead of the CSV file name. To keep an original and edited version of the dataset, the zip file of the dataset simply needs to be unzipped again and kept under a different name from the edited version in the location where all the project files are being held. This shows that the method of "_Programmatically download files_" is the next simplest and easiest method out of all five methods.

**_Dataset variables_**:

*   *_**"population"**_: Populations of cities within countries during 2019*
*   *_**"city"**_: City names*
*   *_**"country"**_: Country names*

The **"population"** column is needed because it contains the city populations during 2019. This population data can be shared with the other dataset because the other dataset has values covering from 2017 to 2021. The populations during 2019 are within the time range. The population data is also needed to answer the research question since the research question requires information about the city populations.

The **"city"** and **"country"** columns are needed because the data in the columns will be used to merge the two datasets together.

The rest of the columns containing the _city ASCII string_, _city latitude_, _city longitude_, _alpha-2 country code_, _alpha-3 country code_, _city admin name_, and country's _capital_ are not needed to answer the research question.

In [404]:
#FILL IN 2nd data gathering and loading method
import zipfile

with zipfile.ZipFile("worldcities.zip", "r") as zip_ref: # opens zip file in read mode
    zip_ref.extractall("worldcities/") # extracts all data in zip file

Optional data storing step: You may save your raw dataset files to the local data store before moving to the next step.

In [405]:
#Optional: store the raw data in your local data store
df_world_cities_dataset = pd.read_csv("worldcities/worldcities.csv") # loads CSV file of second dataset
df_world_cities_dataset # displays second dataset

city   city_ascii      lat       lng        country iso2 iso3  \
0            Tokyo        Tokyo  35.6850  139.7514          Japan   JP  JPN   
1         New York     New York  40.6943  -73.9249  United States   US  USA   
2      Mexico City  Mexico City  19.4424  -99.1310         Mexico   MX  MEX   
3           Mumbai       Mumbai  19.0170   72.8570          India   IN  IND   
4        São Paulo    Sao Paulo -23.5587  -46.6250         Brazil   BR  BRA   
...            ...          ...      ...       ...            ...  ...  ...   
15488  Timmiarmiut  Timmiarmiut  62.5333  -42.2167      Greenland   GL  GRL   
15489  Cheremoshna  Cheremoshna  51.3894   30.0989        Ukraine   UA  UKR   
15490    Ambarchik    Ambarchik  69.6510  162.3336         Russia   RU  RUS   
15491      Nordvik      Nordvik  74.0165  111.5100         Russia   RU  RUS   
15492      Ennadai      Ennadai  61.1333 -100.8833         Canada   CA  CAN   

               admin_name  capital  population          id  
0                   Tōkyō  primary  35676000.0  1392685764  
1                New York      NaN  19354922.0  1840034016  
2        Ciudad de México  primary  19028000.0  1484247881  
3             Mahārāshtra    admin  18978000.0  1356226629  
4               São Paulo    admin  18845000.0  1076532519  
...                   ...      ...         ...         ...  
15488            Kujalleq      NaN        10.0  1304206491  
15489   Kyyivs’ka Oblast’      NaN         0.0  1804043438  
15490    Sakha (Yakutiya)      NaN         0.0  1643739159  
15491  Krasnoyarskiy Kray      NaN         0.0  1643587468  
15492             Nunavut      NaN         0.0  1124019423  

[15493 rows x 11 columns]

## 2. Assess data

Assess the data according to data quality and tidiness metrics using the report below.

List **two** data quality issues and **two** tidiness issues. Assess each data issue visually **and** programmatically, then briefly describe the issue you find.  **Make sure you include justifications for the methods you use for the assessment.**

### Quality Issue 1:

#### First Dataset

In [406]:
#FILL IN - Inspecting the dataframe visually
df_polluted_cities_and_countries.iloc[[22, 32, 50, 109, 170, 179, 201, 210, 311, 410, 606, 1568, 4673]] # displays rows 23, 33, 51, 110, 171, 180, 202, 211, 312, 411, 607, 1569, and 4674 of first dataset

Rank                         City  2021 JAN(2021) FEB(2021) MAR(2021)  \
22      23            Sahiwal, Pakistan  81.7      83.7     116.2      83.1   
32      33                Amroha, India  75.0         -         -     105.1   
50      51       Siddharthanagar, Nepal  64.7     149.6     151.7      80.7   
109    110                 Angol, Chile  47.7       7.2         7       3.2   
170    171                 Yanta, China  43.2     108.2         -         -   
179    180  Laktasi, Bosnia Herzegovina  42.8      72.8      66.7      47.2   
201    202              Qingnian, China  42.0     122.2         -         -   
210    211                  Juye, China  41.1         -         -         -   
311    312             Kosjeric, Serbia  35.5      55.5      48.5      40.7   
410    411                Uzice, Serbia  32.1      46.7      43.2      40.4   
606    607            Mamponteng, Ghana  26.5      41.2      31.1      26.7   
1568  1569               Koreatown, USA  16.2      13.3      12.9       2.6   
4673  4674               Quito, Ecuador   8.4       8.8       7.3       7.9   

     APR(2021) MAY(2021) JUN(2021) JUL(2021) AUG(2021) SEP(2021) OCT(2021)  \
22        58.4      50.1      50.3      53.9      62.9        71     104.4   
32        76.9      58.3      48.5      41.6      41.3        36      89.6   
50          85      32.4      23.5      18.8        21      22.7         -   
109          -      56.2     125.3      89.9      67.8      52.9      22.1   
170       40.4      44.1      33.8        23      29.5      24.6      54.7   
179       37.3      17.1      22.4         -      15.3      21.2      50.1   
201         41      42.3      31.7      23.6      28.8      22.7      49.6   
210       40.8      40.1      27.3      17.4        24      19.6      52.7   
311       25.9        16        13         -      11.6      14.7      33.4   
410       24.3       8.1      12.8      13.4       7.6         -      35.7   
606       26.9      22.3         -      25.1      14.4      18.4      19.5   
1568         -         -         -      12.7      15.4      17.8        14   
4673       7.4       6.8       7.6       5.2         -         -        11   

     NOV(2021) DEC(2021)  2020  2019 2018 2017  
22         275         -     -     -    -    -  
32       156.3       125     -     -    -    -  
50           -      81.8     -     -    -    -  
109        8.8        12     -     -    -    -  
170       62.8      58.7     -     -    -    -  
179       57.3      53.4     -     -    -    -  
201       61.5      63.7     -     -    -    -  
210       65.7      70.5     -     -    -    -  
311       53.7      55.7  36.7  28.8    -    -  
410       62.6      57.3  31.7     -    -    -  
606       23.4      38.1     -     -    -    -  
1568      23.3      17.5     -     -    -    -  
4673       8.4      14.6   7.6   8.6    -    -

In [407]:
#FILL IN - Inspecting the dataframe programmatically

df_polluted_cities_and_countries_with_null_values = df_polluted_cities_and_countries.replace('-', np.nan, regex=False) # replaces '-' with NaN (MUST USE "regex=False" or names with '-' in "City" column will be changed into null values) ("regex=False" makes case-insensitive) (MUST USE "inplace=True" TO ENSURE DATA FRAME IS MODIFIED)
df_polluted_cities_and_countries_with_null_values.isnull().sum() # displays amount of null values in each column of "df_polluted_cities_and_countries" dataframe

Rank            0
City            0
2021            0
JAN(2021)     671
FEB(2021)     574
MAR(2021)     193
APR(2021)     256
MAY(2021)      16
JUN(2021)      16
JUL(2021)      22
AUG(2021)       9
SEP(2021)      17
OCT(2021)      26
NOV(2021)      43
DEC(2021)      87
2020         2207
2019         2422
2018         3999
2017         4194
dtype: int64

**_Issue:_** The following columns in the first dataset have missing values: **"JAN(2021)"**, **"FEB(2021)"**, **"MAR(2021)"**, **"APR(2021)"**, **"MAY(2021)"**, **"JUN(2021)"**, **"JUL(2021)"**, **"AUG(2021)"**, **"SEP(2021)"**, **"OCT(2021)"**, **"NOV(2021)"**, **"DEC(2021)"**, **"2020"**, **"2019"**, **"2018"**, and **"2017"**.

The missing data in the **"2019"** column will leave some of the cities' pollution amounts unknown. This will possibly cause some of the most polluted cities to be unaccounted for when answering the research question.

The missing data in the following columns will not affect the analysis for the research question because they are not needed for answering the research question: **"JAN(2021)"**, **"FEB(2021)"**, **"MAR(2021)"**, **"APR(2021)"**, **"MAY(2021)"**, **"JUN(2021)"**, **"JUL(2021)"**, **"AUG(2021)"**, **"SEP(2021)"**, **"OCT(2021)"**, **"NOV(2021)"**, **"DEC(2021)"**, **"2020"**, **"2018"**, and **"2017"**.

**_Justification for Visual Method:_** The purpose of the visual method is to display raw data from rows _23_, _33_, _51_, _110_, _171_, _180_, _202_, _211_, _312_, _411_, _607_, _1569_, and _4674_ to show that there are missing values in the following columns: **"JAN(2021)"**, **"FEB(2021)"**, **"MAR(2021)"**, **"APR(2021)"**, **"MAY(2021)"**, **"JUN(2021)"**, **"JUL(2021)"**, **"AUG(2021)"**, **"SEP(2021)"**, **"OCT(2021)"**, **"NOV(2021)"**, **"DEC(2021)"**, **"2020"**, **"2018"**, and **"2017"**.

**_Justification for Programmatic Method:_** The purpose of the programmatic method is to find and display the amount of missing values in each column of the first dataset.

#### Second Dataset

In [408]:
#FILL IN - Inspecting the dataframe visually
df_world_cities_dataset.head() # displays first 5 rows of raw data

city   city_ascii      lat       lng        country iso2 iso3  \
0        Tokyo        Tokyo  35.6850  139.7514          Japan   JP  JPN   
1     New York     New York  40.6943  -73.9249  United States   US  USA   
2  Mexico City  Mexico City  19.4424  -99.1310         Mexico   MX  MEX   
3       Mumbai       Mumbai  19.0170   72.8570          India   IN  IND   
4    São Paulo    Sao Paulo -23.5587  -46.6250         Brazil   BR  BRA   

         admin_name  capital  population          id  
0             Tōkyō  primary  35676000.0  1392685764  
1          New York      NaN  19354922.0  1840034016  
2  Ciudad de México  primary  19028000.0  1484247881  
3       Mahārāshtra    admin  18978000.0  1356226629  
4         São Paulo    admin  18845000.0  1076532519

In [409]:
#FILL IN - Inspecting the dataframe programmatically
df_world_cities_dataset[['city', 'city_ascii']] # displays raw data in "city" and "city_ascii" columns of "df_polluted_cities_and_countries" dataframe

city   city_ascii
0            Tokyo        Tokyo
1         New York     New York
2      Mexico City  Mexico City
3           Mumbai       Mumbai
4        São Paulo    Sao Paulo
...            ...          ...
15488  Timmiarmiut  Timmiarmiut
15489  Cheremoshna  Cheremoshna
15490    Ambarchik    Ambarchik
15491      Nordvik      Nordvik
15492      Ennadai      Ennadai

[15493 rows x 2 columns]

**_Issue:_** The column called **"city_ascii"** is a duplicate column of the **"city"** column because it contains the same data as the **"city"** column.
	
The unnecessary data from the **"city_ascii"** column will take up space in the dataset and increase the amount of data to analyze through. However, it will not affect the analysis for the research question because the data in the **"city_ascii"** column is the exact same as the data in the **"city"** column. Therefore, it will not affect the merging of the two datasets and the analysis.

**_Justification for Visual Method:_** The main purpose of the visual method is to display the first five rows of raw data from the **"city"** and **"city_ascii"** columns to prove that the **"city_ascii"** column is a duplicate column of the **"city"** column.

**_Justification for Programmatic Method:_** The purpose of the programmatic method is to show the raw data from the **"city"** and **"city_ascii"** columns to prove that the **"city_ascii"** column is a duplicate column of the **"city"** column.

### Quality Issue 2:

#### First Dataset

In [410]:
#FILL IN - Inspecting the dataframe visually
df_polluted_cities_and_countries.head() # displays first 5 rows of raw data

Rank              City   2021 JAN(2021) FEB(2021) MAR(2021) APR(2021)  \
0     1    Bhiwadi, India  106.2     145.8     129.8     120.2     125.7   
1     2  Ghaziabad, India  102.0     199.9     172.2      97.8      86.3   
2     3      Hotan, China  101.5         -         -       158      91.1   
3     4      Delhi, India   96.4     183.7     142.2      80.5      72.9   
4     5    Jaunpur, India   95.3     182.2     143.5        91        70   

  MAY(2021) JUN(2021) JUL(2021) AUG(2021) SEP(2021) OCT(2021) NOV(2021)  \
0      86.5      95.9      55.6      55.4      37.1      91.1     188.6   
1      52.9      47.2      35.3      37.6      30.8      89.7     218.3   
2     167.4      57.4      70.9      93.2      79.3     126.1     111.5   
3      47.4      47.1      35.6      36.9      30.2      73.7     224.1   
4      51.1      40.7      33.5      34.2      36.8      75.7       196   

  DEC(2021)   2020   2019   2018   2017  
0     136.6   95.5   83.4  125.4      -  
1       163  106.6  110.2  135.2  144.6  
2      62.6  110.2  110.1    116   91.9  
3     186.4   84.1   98.6  113.5  108.2  
4     195.7      -      -      -      -

In [411]:
#FILL IN - Inspecting the dataframe programmatically

df_polluted_cities_and_countries.replace('-', 0, inplace=True, regex=False) # replaces '-' with 0 (MUST USE "regex=False" or names with '-' in "City" column will be changed into null values) ("regex=False" makes case-insensitive) (MUST USE "inplace=True" TO ENSURE DATA FRAME IS MODIFIED)

cols_to_add = ['JAN(2021)', 'FEB(2021)', 'MAR(2021)', 'APR(2021)', 'MAY(2021)', 'JUN(2021)', 'JUL(2021)', 'AUG(2021)', 'SEP(2021)', 'OCT(2021)', 'NOV(2021)', 'DEC(2021)'] # series to hold column names of 2021 months
for column in cols_to_add:
    df_polluted_cities_and_countries[column] = df_polluted_cities_and_countries[column].astype(np.float64) # converts 'JAN(2021)', 'FEB(2021)', 'MAR(2021)', 'APR(2021)', 'MAY(2021)', 'JUN(2021)', 'JUL(2021)', 'AUG(2021)', 'SEP(2021)', 'OCT(2021)', 'NOV(2021)', and 'DEC(2021)' columns from "object" to "float64"

combined_sums = df_polluted_cities_and_countries[['JAN(2021)', 'FEB(2021)', 'MAR(2021)', 'APR(2021)', 'MAY(2021)', 'JUN(2021)', 'JUL(2021)', 'AUG(2021)', 'SEP(2021)', 'OCT(2021)', 'NOV(2021)', 'DEC(2021)']].sum(axis=1, skipna=True, numeric_only=False) # finds sum of each dataframe row under 2021 month columns
print("Combined Values of 2021 Months:\n")
print(combined_sums)

print("\n\nColumn 2021 values:")
df_polluted_cities_and_countries['2021']

Combined Values of 2021 Months:

0       1268.3
1       1231.0
2       1017.5
3       1160.7
4       1150.4
         ...  
6470      29.2
6471      24.9
6472      23.5
6473      23.3
6474      18.3
Length: 6475, dtype: float64


Column 2021 values:


0       106.2
1       102.0
2       101.5
3        96.4
4        95.3
        ...  
6470      2.4
6471      2.1
6472      2.0
6473      1.9
6474      1.5
Name: 2021, Length: 6475, dtype: float64

In [412]:
# puts first dataframe back to original state
df_polluted_cities_and_countries.replace(0, '-', inplace=True, regex=False) # replaces 0 with '-' (MUST USE "regex=False" or names with '-' in "City" column will be changed into null values) ("regex=False" makes case-insensitive) (MUST USE "inplace=True" TO ENSURE DATA FRAME IS MODIFIED)

**_Issue:_** In each row, the computation of twelve air quality index values for 2021 under columns, **"JAN(2021)"**, **"FEB(2021)"**, **"MAR(2021)"**, **"APR(2021)"**, **"MAY(2021)"**, **"JUN(2021)"**, **"JUL(2021)"**, **"AUG(2021)"**, **"SEP(2021)"**, **"OCT(2021)"**, **"NOV(2021)"**, and **"DEC(2021)"**, does not match the corresponding total air quality index value for 2021 under the **"2021"** column. This makes the data under the twelve columns inaccurate.
	
The inaccurate data will not affect the analysis for the research question because the twelve columns are not needed for answering the research question.

**_Justification for Visual Method:_** The main purpose of the visual method is to display the first five rows of raw data from the **"JAN(2021)"**, **"FEB(2021)"**, **"MAR(2021)"**, **"APR(2021)"**, **"MAY(2021)"**, **"JUN(2021)"**, **"JUL(2021)"**, **"AUG(2021)"**, **"SEP(2021)"**, **"OCT(2021)"**, **"NOV(2021)"**, **"DEC(2021)"**, and **"2021"** columns to show that the columns contain inaccurate values that do not equal the corresponding total air quality index value for 2021 in the **"2021"** column.

**_Justification for Programmatic Method:_** The main purpose of the programmatic method is to add the twelve air quality index values for 2021 under columns, **"JAN(2021)"**, **"FEB(2021)"**, **"MAR(2021)"**, **"APR(2021)"**, **"MAY(2021)"**, **"JUN(2021)"**, **"JUL(2021)"**, **"AUG(2021)"**, **"SEP(2021)"**, **"OCT(2021)"**, **"NOV(2021)"**, and **"DEC(2021)"**, then display the resulting sums to prove that the sums do not match the corresponding total air quality index values for 2021 in the **"2021"** column.

#### Second Dataset

In [413]:
#FILL IN - Inspecting the dataframe visually
df_world_cities_dataset.iloc[[1, 8, 24, 32, 53, 424, 548, 600, 601, 602, 3780, 4357]] # displays rows 1, 8, 24, 32, 53, 424, 548, 600, 601, 602, 3780, and 4357 of second dataset

city   city_ascii      lat       lng        country iso2 iso3  \
1        New York     New York  40.6943  -73.9249  United States   US  USA   
8     Los Angeles  Los Angeles  34.1139 -118.4068  United States   US  USA   
24        Chicago      Chicago  41.8373  -87.6862  United States   US  USA   
32      Hong Kong    Hong Kong  22.3050  114.1850      Hong Kong   HK  HKG   
53      Singapore    Singapore   1.2930  103.8558      Singapore   SG  SGP   
424        Namp’o       Namp'o  38.7669  125.4524   Korea, North   KP  PRK   
548      Windhoek     Windhoek -22.5700   17.0835        Namibia  NaN  NAM   
600       Al Quds      Al Quds  31.7764   35.2269      West Bank   XW  XWB   
601   Philipsburg  Philipsburg  18.0255  -63.0450   Sint Maarten   SX  SXM   
602        Gitega       Gitega  -3.4271   29.9246        Burundi   BI  BDI   
3780        Rundu        Rundu -17.9200   19.7499        Namibia  NaN  NAM   
4357   Swakopmund   Swakopmund -22.6689   14.5350        Namibia  NaN  NAM   

        admin_name  capital  population          id  
1         New York      NaN  19354922.0  1840034016  
8       California      NaN  12815475.0  1840020491  
24        Illinois      NaN   8675982.0  1840000494  
32             NaN      NaN   7206000.0  1344982653  
53             NaN  primary   5183700.0  1702341327  
424            NaN    admin   1127000.0  1408892768  
548         Khomas  primary    268132.0  1516802003  
600            NaN  primary         NaN  1934000006  
601            NaN  primary         NaN  1534859858  
602         Gitega  primary         NaN  1108778000  
3780  Kavango East    admin     58172.0  1516885937  
4357        Erongo    admin     29491.0  1516214290

In [414]:
#FILL IN - Inspecting the dataframe programmatically
df_world_cities_dataset.isnull().sum() # displays amount of null values in each column of "df_world_cities_dataset" dataframe

city              0
city_ascii        0
lat               0
lng               0
country           0
iso2             31
iso3              0
admin_name      191
capital       10247
population     1685
id                0
dtype: int64

**_Issue:_**  The following columns have missing values: **"iso2"**, **"admin_name"**, **"capital"**, and **"population"**.

The missing values in the **"population"** column will leave some of the cities' population amounts unknown. This will possibly cause some of the most populated cities to be unaccounted for when answering the research question.

The missing values in the **"admin_name"** and **"capital"** columns will not affect the analysis for the research question because the data in those columns are not needed to answer the research question.

**_Justification for Visual Method:_** The main purpose of the visual method is to display raw data from rows _1_, _8_, _24_, _32_, _53_, _424_, _548_, _600_, _601_, _602_, _3780_, and _4357_ to show that there are missing values in the following columns: **"iso2"**, **"admin_name"**, **"capital"**, and **"population"**.

**_Justification for Programmatic Method:_** The main purpose of the visual method is to find and display the amount of null values in each column of the second dataset.

### Tidiness Issue 1:

#### First Dataset

In [415]:
#FILL IN - Inspecting the dataframe visually
df_polluted_cities_and_countries.head() # displays first 5 rows of raw data

Rank              City   2021 JAN(2021) FEB(2021) MAR(2021) APR(2021)  \
0     1    Bhiwadi, India  106.2     145.8     129.8     120.2     125.7   
1     2  Ghaziabad, India  102.0     199.9     172.2      97.8      86.3   
2     3      Hotan, China  101.5         -         -     158.0      91.1   
3     4      Delhi, India   96.4     183.7     142.2      80.5      72.9   
4     5    Jaunpur, India   95.3     182.2     143.5      91.0      70.0   

  MAY(2021) JUN(2021) JUL(2021) AUG(2021) SEP(2021) OCT(2021) NOV(2021)  \
0      86.5      95.9      55.6      55.4      37.1      91.1     188.6   
1      52.9      47.2      35.3      37.6      30.8      89.7     218.3   
2     167.4      57.4      70.9      93.2      79.3     126.1     111.5   
3      47.4      47.1      35.6      36.9      30.2      73.7     224.1   
4      51.1      40.7      33.5      34.2      36.8      75.7     196.0   

  DEC(2021)   2020   2019   2018   2017  
0     136.6   95.5   83.4  125.4      -  
1     163.0  106.6  110.2  135.2  144.6  
2      62.6  110.2  110.1    116   91.9  
3     186.4   84.1   98.6  113.5  108.2  
4     195.7      -      -      -      -

In [416]:
#FILL IN - Inspecting the dataframe programmatically
df_polluted_cities_and_countries[["City"]] # displays raw data in "City" column of "df_polluted_cities_and_countries" dataframe

City
0            Bhiwadi, India
1          Ghaziabad, India
2              Hotan, China
3              Delhi, India
4            Jaunpur, India
...                     ...
6470  Mornington, Australia
6471   Emu River, Australia
6472     Judbury, Australia
6473   St Helens, Australia
6474        Chu, Kazakhstan

[6475 rows x 1 columns]

**_Issue:_** The column named **"City"** contains both the _country_ and **city** names when it should only contain the _country_ names. 

This violates the _first data tidiness rule_: **Each variable forms a column**. 

The column will prevent the first and second datasets from being merged together because the **"City"** must only contain **city** names in order to merge the two datasets.

**_Justification for Visual Method:_** The main purpose of the visual method is to display the first five rows of raw data from the **"City"** column to show the column contains both the _country_ and **city** names when it should only contain the _country_ names.

**_Justification for Programmatic Method:_** The main purpose of the programmatic method is to display data in **"City"** column to show the column contains both the _country_ and **city** names when it should only contain the _country_ names.

#### Second Dataset

In [417]:
#FILL IN - Inspecting the dataframe visually
df_world_cities_dataset.head() # displays first 5 rows of raw data

city   city_ascii      lat       lng        country iso2 iso3  \
0        Tokyo        Tokyo  35.6850  139.7514          Japan   JP  JPN   
1     New York     New York  40.6943  -73.9249  United States   US  USA   
2  Mexico City  Mexico City  19.4424  -99.1310         Mexico   MX  MEX   
3       Mumbai       Mumbai  19.0170   72.8570          India   IN  IND   
4    São Paulo    Sao Paulo -23.5587  -46.6250         Brazil   BR  BRA   

         admin_name  capital  population          id  
0             Tōkyō  primary  35676000.0  1392685764  
1          New York      NaN  19354922.0  1840034016  
2  Ciudad de México  primary  19028000.0  1484247881  
3       Mahārāshtra    admin  18978000.0  1356226629  
4         São Paulo    admin  18845000.0  1076532519

In [423]:
#FILL IN - Inspecting the dataframe programmatically
df_world_cities_dataset.head(10) # displays column names and their corresponding data for first 10 rows of "df_world_cities_dataset" dataframe

city   city_ascii      lat       lng        country iso2 iso3  \
0        Tokyo        Tokyo  35.6850  139.7514          Japan   JP  JPN   
1     New York     New York  40.6943  -73.9249  United States   US  USA   
2  Mexico City  Mexico City  19.4424  -99.1310         Mexico   MX  MEX   
3       Mumbai       Mumbai  19.0170   72.8570          India   IN  IND   
4    São Paulo    Sao Paulo -23.5587  -46.6250         Brazil   BR  BRA   
5        Delhi        Delhi  28.6700   77.2300          India   IN  IND   
6     Shanghai     Shanghai  31.2165  121.4365          China   CN  CHN   
7      Kolkata      Kolkata  22.4950   88.3247          India   IN  IND   
8  Los Angeles  Los Angeles  34.1139 -118.4068  United States   US  USA   
9        Dhaka        Dhaka  23.7231   90.4086     Bangladesh   BD  BGD   

         admin_name  capital  population          id  
0             Tōkyō  primary  35676000.0  1392685764  
1          New York      NaN  19354922.0  1840034016  
2  Ciudad de México  primary  19028000.0  1484247881  
3       Mahārāshtra    admin  18978000.0  1356226629  
4         São Paulo    admin  18845000.0  1076532519  
5             Delhi    admin  15926000.0  1356872604  
6          Shanghai    admin  14987000.0  1156073548  
7       West Bengal    admin  14787000.0  1356060520  
8        California      NaN  12815475.0  1840020491  
9             Dhaka  primary  12797394.0  1050529279

**_Issue:_** The **"capital"** column should be in a separate table because it is not a demographic detail of the cities like the rest of the columns in the dataset. 

This violates the _third data tidiness rule_: **Each type of observational unit forms a table**.

The contrasting data will not affect the analysis for the research question because it is not needed for answering the research question.

**_Justification for Visual Method:_** The main purpose of the visual method is to display all of the column names and their corresponding data for five rows of the second dataset to show that the **"capital"** column should be in a separate table because it is not a demographic detail of the cities like the rest of the columns in the dataset. 

**_Justification for Programmatic Method:_** The main purpose of the visual method is to display all of the column names and their corresponding data of the second dataset to show that the **"capital"** column should be in a separate table because it is not a demographic detail of the cities like the rest of the columns in the dataset.

### Tidiness Issue 2: 

#### First Dataset

In [419]:
#FILL IN - Inspecting the dataframe visually
df_polluted_cities_and_countries.head() # displays first 5 rows of raw data

Rank              City   2021 JAN(2021) FEB(2021) MAR(2021) APR(2021)  \
0     1    Bhiwadi, India  106.2     145.8     129.8     120.2     125.7   
1     2  Ghaziabad, India  102.0     199.9     172.2      97.8      86.3   
2     3      Hotan, China  101.5         -         -     158.0      91.1   
3     4      Delhi, India   96.4     183.7     142.2      80.5      72.9   
4     5    Jaunpur, India   95.3     182.2     143.5      91.0      70.0   

  MAY(2021) JUN(2021) JUL(2021) AUG(2021) SEP(2021) OCT(2021) NOV(2021)  \
0      86.5      95.9      55.6      55.4      37.1      91.1     188.6   
1      52.9      47.2      35.3      37.6      30.8      89.7     218.3   
2     167.4      57.4      70.9      93.2      79.3     126.1     111.5   
3      47.4      47.1      35.6      36.9      30.2      73.7     224.1   
4      51.1      40.7      33.5      34.2      36.8      75.7     196.0   

  DEC(2021)   2020   2019   2018   2017  
0     136.6   95.5   83.4  125.4      -  
1     163.0  106.6  110.2  135.2  144.6  
2      62.6  110.2  110.1    116   91.9  
3     186.4   84.1   98.6  113.5  108.2  
4     195.7      -      -      -      -

In [420]:
#FILL IN - Inspecting the dataframe programmatically
df_polluted_cities_and_countries.columns # displays column names of "df_polluted_cities_and_countries" dataframe

Index(['Rank', 'City', '2021', 'JAN(2021)', 'FEB(2021)', 'MAR(2021)',
       'APR(2021)', 'MAY(2021)', 'JUN(2021)', 'JUL(2021)', 'AUG(2021)',
       'SEP(2021)', 'OCT(2021)', 'NOV(2021)', 'DEC(2021)', '2020', '2019',
       '2018', '2017'],
      dtype='object')

**_Issue:_** The data in the following columns should be in a separate table because they contain air quality index values for each month of year 2021 instead of one whole year like the other columns holding air quality index values: **"JAN(2021)"**, **"FEB(2021)"**, **"MAR(2021)"**, **"APR(2021)"**, **"MAY(2021)"**, **"JUN(2021)"**, **"JUL(2021)"**, **"AUG(2021)"**, **"SEP(2021)"**, **"OCT(2021)"**, **"NOV(2021)"**, and **"DEC(2021)"**.

This violates the _third data tidiness rule_: **Each type of observational unit forms a table**.
	
The contrasting data will not affect the analysis for the research question because it is not needed for answering the research question.

**_Justification for Visual Method:_** The main purpose of the visual method is to display all of the column names of the first dataset to show that the **"JAN(2021)"**, **"FEB(2021)"**, **"MAR(2021)"**, **"APR(2021)"**, **"MAY(2021)"**, **"JUN(2021)"**, **"JUL(2021)"**, **"AUG(2021)"**, **"SEP(2021)"**, **"OCT(2021)"**, **"NOV(2021)"**, and **"DEC(2021)"** columns hold data from the months during 2021, whereas the columns **"2021"**, **"2020"**, **"2019"**, **"2018"**, and **"2017"** hold data for their column year.

**_Justification for Programmatic Method:_** The main purpose of the programmatic method is to display all of the column names of the first dataset to show that the **"JAN(2021)"**, **"FEB(2021)"**, **"MAR(2021)"**, **"APR(2021)"**, **"MAY(2021)"**, **"JUN(2021)"**, **"JUL(2021)"**, **"AUG(2021)"**, **"SEP(2021)"**, **"OCT(2021)"**, **"NOV(2021)"**, and **"DEC(2021)"** columns are the only columns that hold data from the months during 2021, whereas the columns **"2021"**, **"2020"**, **"2019"**, **"2018"**, and **"2017"** hold data for their column year.

#### Second Dataset

In [ ]:
#FILL IN - Inspecting the dataframe visually

In [ ]:
#FILL IN - Inspecting the dataframe programmatically

Issue and justification: *FILL IN*

## 3. Clean data
Clean the data to solve the 4 issues corresponding to data quality and tidiness found in the assessing step. **Make sure you include justifications for your cleaning decisions.**

After the cleaning for each issue, please use **either** the visually or programatical method to validate the cleaning was succesful.

At this stage, you are also expected to remove variables that are unnecessary for your analysis and combine your datasets. Depending on your datasets, you may choose to perform variable combination and elimination before or after the cleaning stage. Your dataset must have **at least** 4 variables after combining the data.

In [ ]:
# FILL IN - Make copies of the datasets to ensure the raw dataframes 
# are not impacted

### **Quality Issue 1: FILL IN**

In [ ]:
# FILL IN - Apply the cleaning strategy

In [ ]:
# FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Quality Issue 2: FILL IN**

In [ ]:
#FILL IN - Apply the cleaning strategy

In [ ]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Tidiness Issue 1: FILL IN**

In [ ]:
#FILL IN - Apply the cleaning strategy

In [ ]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Tidiness Issue 2: FILL IN**

In [ ]:
#FILL IN - Apply the cleaning strategy

In [ ]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Remove unnecessary variables and combine datasets**

Depending on the datasets, you can also peform the combination before the cleaning steps.

In [ ]:
#FILL IN - Remove unnecessary variables and combine datasets

## 4. Update your data store
Update your local database/data store with the cleaned data, following best practices for storing your cleaned data:

- Must maintain different instances / versions of data (raw and cleaned data)
- Must name the dataset files informatively
- Ensure both the raw and cleaned data is saved to your database/data store

In [ ]:
#FILL IN - saving data

## 5. Answer the research question

### **5.1:** Define and answer the research question 
Going back to the problem statement in step 1, use the cleaned data to answer the question you raised. Produce **at least** two visualizations using the cleaned data and explain how they help you answer the question.

*Research question:* FILL IN from answer to Step 1

In [ ]:
#Visual 1 - FILL IN

*Answer to research question:* FILL IN

In [ ]:
#Visual 2 - FILL IN

*Answer to research question:* FILL IN

### **5.2:** Reflection
In 2-4 sentences, if you had more time to complete the project, what actions would you take? For example, which data quality and structural issues would you look into further, and what research questions would you further explore?

*Answer:* FILL IN